File: src/pingui/monitor/route_history.py
"""Last-known hop IP tracking for inactive route display."""

from __future__ import annotations

from pingui.models import TIMEOUT_IP, HopNode


def record_last_known(
    last_known: dict[int, HopNode],
    nodes: list[HopNode],
) -> None:
    """Remember the latest reachable IP observed at each hop index."""
    for node in nodes:
        if node.is_timeout or node.ip == TIMEOUT_IP:
            continue
        last_known[node.hop] = node


def route_with_last_known_ips(
    route: list[HopNode],
    last_known: dict[int, HopNode],
) -> list[HopNode]:
    """Replace timeout hops with last known IPs for inactive chain display."""
    result: list[HopNode] = []
    for node in route:
        if not node.is_timeout and node.ip != TIMEOUT_IP:
            result.append(node)
            continue
        known = last_known.get(node.hop)
        if known is None:
            result.append(node)
            continue
        result.append(
            HopNode(
                hop=node.hop,
                ip=known.ip,
                ping_ms=known.ping_ms,
                is_timeout=False,
            )
        )
    return result

File: java/src/main/java/io/pingui/monitor/RouteHistory.java
package io.pingui.monitor;

import io.pingui.model.Models.HopNode;
import java.util.ArrayList;
import java.util.List;
import java.util.Map;

/** Last-known hop IP tracking for inactive route display. */
public final class RouteHistory {
    private RouteHistory() {}

    public static void recordLastKnown(Map<Integer, HopNode> lastKnown, List<HopNode> nodes) {
        for (HopNode node : nodes) {
            if (node.isReachable()) {
                lastKnown.put(node.hop(), node);
            }
        }
    }

    public static List<HopNode> routeWithLastKnownIps(List<HopNode> route, Map<Integer, HopNode> lastKnown) {
        List<HopNode> result = new ArrayList<>();
        for (HopNode node : route) {
            if (node.isReachable()) {
                result.add(node);
                continue;
            }
            HopNode known = lastKnown.get(node.hop());
            if (known == null) {
                result.add(node);
                continue;
            }
            result.add(new HopNode(node.hop(), known.ip(), known.pingMs(), false));
        }
        return List.copyOf(result);
    }
}
